ctapipe/reco/shower_reconstructor.py

```diff
@@ -0,0 +1,169 @@
+        super().__init__(subarray=subarray, **kwargs)
+
+    def __call__(self, event):
+        if self.geometry_seed not in event.dl2.stereo.geometry:
+            raise ValueError()
+
+        # get all telescope properties that we need for the ShowermodelPredictor
+        tel_positions = {}
+        tel_solid_angles = {}
+        tel_mirror_area = {}
+        for tel_id in event.dl1.tel.keys():
+            tel_positions[tel_id] = self.subarray.positions[tel_id]
+            geometry = self.subarray.tel[tel_id].camera.geometry
+            tel_solid_angles[tel_id] = geometry.transform_to(TelescopeFrame()).pix_area
+            tel_mirror_area[tel_id] = self.subarray.tel[tel_id].optics.mirror_area
+
+        self.tel_positions = tel_positions
+        self.tel_solid_angles = tel_solid_angles
+        self.tel_mirror_area = tel_mirror_area
+        self.tel_pix_coords_altaz = self._tel_pix_coords_altaz(event)
+
+        # for likelihood
+        self.DL1CamContainers = event.dl1.tel.items()
+
+        shower_parameters, errors = self._fit(event)
+
+        event.dl2.stereo.geometry[
+            self.__class__.__name__
+        ] = ReconstructedGeometryContainer(
+            total_photons=shower_parameters["total_photons"],
+            total_photons_uncert=errors["total_photons"],
+            core_x=shower_parameters["x"],
+            core_y=shower_parameters["y"],
+            core_uncert_x=errors["x"],
+            core_uncert_y=errors["y"],
+            alt=shower_parameters["altitude"],
+            az=shower_parameters["azimuth"],
+            alt_uncert=errors["altitude"],
+            az_uncert=errors["azimuth"],
+            h_max=shower_parameters["h_max"],
+            h_max_uncert=errors["h_max"],
+            width=shower_parameters["width"],
+            width_uncert=errors["width"],
+            length=shower_parameters["length"],
+            length_uncert=errors["length"],
+            telescopes=list(event.dl1.tel.keys()),
+            is_valid=True,
+            prefix=self.__class__.__name__,
+        )
+
+    def _fit(self, event):
+        # just call iminiut with likelihood and some seeds
+        # may also need limits
+        seeds = self._seeds(event)
+        minimizer = Minuit(self._likelihood, **seeds)
+
+        minimizer.limits["total_photons"] = [0, None]
+        minimizer.limits["x"] = [None, None]
+        minimizer.limits["y"] = [None, None]
+        minimizer.limits["altitude"] = [0, 90]
+        minimizer.limits["azimuth"] = [0, 360]
+        minimizer.limits["h_max"] = [0, None]
+        minimizer.limits["width"] = [1e-8, None]
+        minimizer.limits["length"] = [1e-8, None]
+
+        minimizer.errordef = Minuit.LIKELIHOOD
+        minimizer.migrad()
+        fit = minimizer.values
+        fit_errors = minimizer.errors
+
+        return fit, fit_errors
+
+    def _likelihood(self, total_photons, x, y, azimuth, altitude, h_max, width, length):
+
+        model = GaussianShowermodel(
```

Comment:
This also seems very expensive, especially the adding of astropy units inside the likelihood.

Reply:
Yes, astropy units are the biggest time consuming thing right now (I did profiling). I guess I have drop all units in the model and in the predictor.